cbpv.zy

```diff
@@ -8,6 +8,31 @@ data Exception where
   | FreeVariable()
 end
 
+codata Alg (M: VType -> CType) (B: CType) where
+  | .bindAlg(): forall (A: VType) . U (M A) -> U (A -> B) -> B
+end
+
+def fn bind-arrow (M: VType -> CType) (B: CType) (alg: U (Alg M B)) (A: VType) : Alg M (A -> B) =
+  comatch
+  | .bindAlg() -> fn A' m f -> ! alg .bindAlg() @(A') m { fn a' -> ! f a' }
```

Comment:
Couldn't this just be 
```
  | .bindAlg() -> ! alg .bindAlg()
```

Reply:
Does this type check?

Reply:
> Couldn't this just be
> 
> ```
>   | .bindAlg() -> ! alg .bindAlg()
> ```

They are, but shouldn't. I made another attempt, which added another argument so that it should work, but the current Zydeco type checker is still refusing. I believe that's because of a bug in the current implementation regarding higher-kinded types.

Reply:
Pushed to [here](https://github.com/zydeco-lang/monadic-interpreter/pull/1/files/ae492e90d05a4aac5ad02d52610c45997d0e3cdc..a51c396f1edbfe1c8f747c7f4b0c2414ddce4859).